extensions/amp-viqeo-player/0.1/amp-viqeo-player.js

```diff
@@ -0,0 +1,440 @@
+    this.playerReadyResolver_ = null;
+
+    /** @private {?number} */
+    this.volume_ = null;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+
+    /** @private {?Object} */
+    this.viqeoPlayer_ = null;
+
+    /** @private {boolean} */
+    this.kindIsProd_ = true;
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url('https://static.viqeo.tv', opt_onLayout);
+    this.preconnect.url('https://stage.embed.viqeo.tv', opt_onLayout);
+  }
+
+  /**
+   * @param {!Layout} layout
+   * @return {boolean}
+   * @override
+   */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    return null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.videoId_ = user().assert(
+        this.element.getAttribute('data-videoid'),
+        'The data-videoid attribute is required for <amp-viqeo-player> %s',
+        this.element);
+
+    this.profileId_ = user().assert(
+        this.element.getAttribute('data-profileid'),
+        'The data-profileid attribute is required for <amp-viqeo-player> %s',
+        this.element);
+
+    this.kindIsProd_ = this.element.getAttribute('data-kind') !== 'stage';
+
+
+    const deferred = new Deferred();
+    this.playerReadyPromise_ = deferred.promise;
+    this.playerReadyResolver_ = deferred.resolve;
+
+    installVideoManagerForDoc(this.element);
+    Services.videoManagerForDoc(this.element).register(this);
+
+  }
+
+  /** @override */
+  layoutCallback() {
+
+    const iframeStyle = this.element.getAttribute('data-iframe-style')
+        || 'position: absolute';
+    const iframeHeight = this.element.getAttribute('data-iframe-height')
+        || '100%';
+    const iframeWidth = this.element.getAttribute('data-iframe-width')
+        || '100%';
+
+    let scriptPlayerInit = this.element.getAttribute('data-script-url');
+    scriptPlayerInit =
+        (scriptPlayerInit
+            && scriptPlayerInit.length && decodeURI(scriptPlayerInit)
+        )
+        ||
+        (this.kindIsProd_
+          ? 'https://cdn.viqeo.tv/js/vq_player_init.js'
+          : 'https://static.viqeo.tv/js/vq_player_init.js?branch=dev1'
+        );
+
+    const ampIframe = this.element.ownerDocument.createElement('iframe');
+    ampIframe.src = 'about:blank';
+    if (ampIframe.contentWindow) {
+      frameLoaded.call(this);
+    } else {
+      ampIframe.onload = frameLoaded.bind(this);
+    }
+
+    this.element.appendChild(ampIframe);
+    this.iframe_ = ampIframe;
+
+    return this.loadPromise(ampIframe);
+
+    /** @this {AmpViqeoPlayer} */
+    function frameLoaded() {
+      const doc = ampIframe.contentWindow.document;
+      setStyles(doc.body, {
+        'marginLeft': 0,
+        'marginRight': 0,
+        'marginTop': 0,
+        'marginBottom': 0,
+      });
+
+      const scr = doc.createElement('script');
```

Comment:
@sdbaron I mentioned in the previous comment, extensions are not allowed to inject 3P scripts in AMP for obvious security reasons. Only code that has been reviewed by us can run in the top level. We unfortunately can not allow this under any situation.

Reply:
@aghassemi I load script not on the top level, but to the next level iframe that was created:

const ampIframe = this.element.ownerDocument.createElement('iframe');

Am I wrong and need iframe that should be loaded from another domain?

Reply:
@sdbaron 
Please take a look at components such as amp-facebook, amp-bodymovin-animation, amp-twitter on how to have a component that is backed by a 3P iframe. Essentially they are two pieces, one the component file which you already have and 2 a file in out "3p" folder that runs the code for the iframe inside a x-origin sandbox environment. 

`amp-bodymovin-animation` is probably the best example since it is similar to a video player ( it is an animation player). The two important files are: https://github.com/ampproject/amphtml/blob/master/extensions/amp-bodymovin-animation/0.1/amp-bodymovin-animation.js and https://github.com/ampproject/amphtml/blob/master/3p/bodymovinanimation.js (which runs inside the iframe)

Reply:
(`about:blank` run in the same security context as parent, allowing your script to access cookies, etc.. from top level: see https://jsbin.com/yuvemeyusa/2/edit?js,output)